Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp

```diff
@@ -0,0 +1,238 @@
+    }
+
+    void MultiplayerStatSystemComponent::DeclareStatGroup(int uniqueGroupId, const char* groupName)
+    {
+        AZStd::lock_guard lock(m_access);
+        StatGroup* newGroup = m_statGroups.AddNew(uniqueGroupId);
+        newGroup->m_name = groupName;
+    }
+
+    void MultiplayerStatSystemComponent::DeclareStat(int uniqueGroupId, int uniqueStatId, const char* statName)
+    {
+        AZStd::lock_guard lock(m_access);
+        if (StatGroup* group = m_statGroups.Find(uniqueGroupId))
+        {
+            auto* newStat = group->m_stats.AddNew(uniqueStatId);
+            newStat->m_name = statName;
+
+            const auto statIterator = m_statIdToGroupId.find(uniqueStatId);
+            if (statIterator == m_statIdToGroupId.end())
+            {
+                m_statIdToGroupId.insert(AZStd::make_pair(uniqueStatId, uniqueGroupId));
+            }
+            else
+            {
+                AZLOG_WARN("A stat has already been declared using DECLARE_STAT_INT64 with id %d", uniqueStatId);
+            }
+        }
+        else
+        {
+            AZLOG_WARN("Stat group with id %d has not been declared using DECLARE_STAT_GROUP", uniqueGroupId);
+        }
+    }
+
+    void MultiplayerStatSystemComponent::SetStat(int uniqueStatId, double value)
+    {
+        AZStd::lock_guard lock(m_access);
+        const auto statIterator = m_statIdToGroupId.find(uniqueStatId);
+        if (statIterator != m_statIdToGroupId.end())
+        {
+            if (const auto group = m_statGroups.Find(statIterator->second))
+            {
+                if (auto stat = group->m_stats.Find(uniqueStatId))
+                {
+                    stat->m_average.PushEntry(value);
+                }
+                else
+                {
+                    AZLOG_WARN("SET_STAT_UINT64 was called on a stat with id %d of a non-INT64 type", uniqueStatId);
+                }
+                return;
+            }
+        }
+
+        AZLOG_WARN("Stat with id %d has not been declared using DECLARE_STAT_UINT64", uniqueStatId);
+    }
+
+    void MultiplayerStatSystemComponent::RecordMetrics()
+    {
+        if (const auto* eventLoggerFactory = AZ::Interface<AZ::Metrics::IEventLoggerFactory>::Get())
+        {
+            if (auto* eventLogger = eventLoggerFactory->FindEventLogger(NetworkingMetricsId))
+            {
+                AZStd::lock_guard lock(m_access);
+                for (StatGroup& group : m_statGroups.m_items)
+                {
+                    AZStd::vector<AZ::Metrics::EventField> argsContainer;
+
+                    AZStd::for_each(
+                        group.m_stats.m_items.begin(),
+                        group.m_stats.m_items.end(),
+                        [&argsContainer](const CumulativeAverage& stat)
+                        {
+                            argsContainer.emplace_back(stat.m_name.c_str(), stat.m_average.CalculateAverage());
+                        });
```

Comment:
You could use the [from_range](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/std/containers/vector.h#L145-L150) constructor to directly construct the argument container(more info about range 
```suggestion
                     auto TransformAverageToEventArg = [](const CumulativeAverage& stat)
                     {
                         return EventField{ stat.m_name, stat.m_average.CalculateAverage() };
                     };
                    AZStd::vector<AZ::Metrics::EventField> argsContainer(AZStd::from_range, group.m_stats.m_items | AZStd::views::transform(TransformAverageToEventArg));
```

Reply:
I'd rather avoid range API for now. But I'll keep it in mind.